### PR TITLE
Allow `keras.Variable` for loss weights

### LINKED ISSUE #20294: loss_weights depending on epoch number
Hi,
I'm trying to train a multi output nn and I need to change the weight of each loss component depending on the epoch number. In some previous versions of keras I implemented this mechanism by defining each weight of the loss_weights parameter as a K.variable() type, and changing the value with K.set_value() in an on_epoch_begin() method of a custom callback. 
Now, in keras 3.4 this is not allowed, as loss_weights must be a list/dict of float, so within the callback I can't change the value in place with K.set_value(). Is there something I can do to overcome this issue?
Thanks

### PR DESCRIPTION
**Pull request by james77777778**
Fix #20294 

With this PR, we can use `keras.Variable` for loss weights:

```python
import numpy as np

import keras
from keras import layers
from keras import models

x = np.random.rand(8, 3)
y1 = np.random.rand(8, 1)
y2 = np.random.randint(0, 2, (8, 1))

inputs = layers.Input(shape=(3,), name="input_a")
output_a = layers.Dense(1, name="output_a")(inputs)
output_b = layers.Dense(1, name="output_b", activation="sigmoid")(inputs)
model = models.Model(inputs, [output_a, output_b])

loss_weights = [
    keras.Variable(initializer="ones", shape=()),
    keras.Variable(initializer="ones", shape=()),
]
model.compile(
    optimizer="sgd",
    loss=["mean_squared_error", "binary_crossentropy"],
    loss_weights=loss_weights,
)

loss_weights[0].assign(0.00001)
loss_weights[1].assign(1.0)
model.fit(x, (y1, y2), batch_size=2, epochs=1)

# Change the values
loss_weights[0].assign(1.0)
loss_weights[1].assign(0.00001)
model.fit(x, (y1, y2), batch_size=2, epochs=1)


```

```bash
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.7180 - output_a_loss: 6.9707e-06 - output_b_loss: 0.7180  
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 0.6000 - output_a_loss: 0.5999 - output_b_loss: 6.9904e-06
```
We can see that the individual losses correspond to the loss weights.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR! Can you add as a unit test something similar to the code snippet you posted? Once thing I'd like to guard for, is to make sure that the arrays attached to those variables can pass the compilation boundary and can be modified afterwards.

**Comment by james77777778** [comment]
Thanks for pointing that out. It seems the following snippet works fine with tf and torch but fails with jax:

```python
import numpy as np

import keras
from keras import callbacks
from keras import layers
from keras import models

x = np.random.rand(8, 3)
y1 = np.random.rand(8, 1)
y2 = np.random.randint(0, 2, (8, 1))

inputs = layers.Input(shape=(3,), name="input_a")
output_a = layers.Dense(1, name="output_a")(inputs)
output_b = layers.Dense(1, name="output_b", activation="sigmoid")(inputs)
model = models.Model(inputs, [output_a, output_b])


class LossWeightsAdjuster(callbacks.Callback):
    def __init__(self, loss_weights):
        self.a = loss_weights[0]
        self.b = loss_weights[1]

    def on_epoch_begin(self, epoch, logs):
        self.a.assign(float(epoch % 2))
        self.b.assign(float((epoch + 1) % 2))


loss_weights = [
    keras.Variable(initializer="ones", shape=()),
    keras.Variable(initializer="ones", shape=()),
]
loss_weights_adjuster = LossWeightsAdjuster(loss_weights)
model.compile(
    optimizer="sgd",
    loss=["mean_squared_error", "binary_crossentropy"],
    loss_weights=loss_weights,
)
hist = model.fit(
    x, (y1, y2), batch_size=2, epochs=2, callbacks=[loss_weights_adjuster]
)
output_a_loss = hist.history["output_a_loss"]
output_b_loss = hist.history["output_b_loss"]
np.testing.assert_allclose(output_a_loss[0], 0.0)
np.testing.assert_allclose(output_b_loss[1], 0.0)

```

The root cause might be that the states in callbacks doesn't propagate to `*_step` in `JAXTrainer`.
Should we support this feature? I may need to add some codes in `JAXTrainer`.

**Comment by james77777778** [comment]
> I think this is too difficult and messy to support in JAX.
> 
> What I would suggest for this workflow is to update the weights in Python (as floats) and recompile the model when you want the new weight values to be taken into account.

Got it. This PR should be closed now.
